Feat/filter labeled addresses

MStreet3 opened the pull request:
<!--- Does this work have a corresponding ticket?  Please link it here as well as one of:
    - the PR title
    - branch name
    - commit message
[LINK-777](https://smartcontract-it.atlassian.net/browse/LINK-777)
--> 

PR #16148 added labels to contract addresses on deployment, but did not introduce filtering.  This introduced a bug where certain changesets in `chainlink-deployments` that relied on the assumption of a single contract per chain for MCMS contracts inadvertently generated proposals with the wrong MCMS contract addresses.

This PR fixes the issue by filtering contracts by labels when loading the address book in a backwards compatible way.  If the passed label set is empty, then only un-labeled contracts are returned.  If there are labels passed in the request, then only contracts that have all labels are returned.

### Requires
<!--- Does this work depend on other open PRs? Please list them.
- https://github.com/smartcontractkit/chainlink-common/pull/7777777
-->

### Supports
<!--- Does this work support other open PRs?  Please list them.
- https://github.com/smartcontractkit/ccip/pull/7777777
-->

MStreet3 committed:
feat(deployment): adds LabeledAddresses type

MStreet3 committed:
chore(deployment): filters addresses by labels

MStreet3 commented on deployment/common/changeset/state_test.go:
```diff
@@ -0,0 +1,203 @@
+			skip:  true,
+			chain: defaultChain,
+
+			// addresses has two proposer contract candidates, which should make it into the final state?
```
@AnieeG please advise on expected behavior. Will fix in this PR.

AnieeG replied:
Based on the switch case flow if `ProposerManyChainMultisig` is found it will just return the contract with type `ProposerManyChainMultisig` first right? That's why I have put the `ManyChainMultisig` at the end with the labels. If it does not explicitly define `ProposerManyChainMultisig` ,`CancellerManyChainMultisig` , `BypasserManyChainMultisig` only then it should go to `ManyChainMultisig`block, unless I am missing something here.

AnieeG replied:
Do you think it's better to call that out in comments or enforce that behavior apart from relying on the switch case flow?

MStreet3 replied:
If There are two possible contracts and only one slot to fill, then The switch doesn't resolve the issue. 

A map is not an ordered set, so it is arbitrary which one contract picked last via the switch statement to fill the specific role. 

Both the typed contract and the labeled contract will pass through the switch statement. But the order that happens is arbitrary because the addresses are a map. 

The unit test that shows this has both a contract of type `ProposerManyChainMultiSig` and `ManyChainMultiSig` with the label `proposer`. 

Which address should be the one in the returned state for `ProposerMcms`?

AnieeG replied:
I totally missed it's a map.
 I think the behavior should be -
if explicitly `ProposerManyChainMultiSig` type is mentioned, it should be selected for `ProposerMcms`. Ideal way would be to only manage it with labels but I am not sure if that would break already existing migration compatibility. For now to me it makes sense to override `ManyChainMultiSig` with the label `proposer` with `ProposerManyChainMultiSig` Or we can throw an error if both are provided in addresses

AnieeG commented on deployment/common/changeset/state.go:
```diff
@@ -79,78 +79,90 @@ func MaybeLoadMCMSWithTimelockState(env deployment.Environment, chainSelectors [
+		multichain = deployment.NewTypeAndVersion(types.ManyChainMultisig, deployment.Version1_0_0)
+
+		// Convert map keys to a slice
+		wantTypes = []deployment.TypeAndVersion{timelock, proposer, canceller, bypasser, callProxy}
```
I think we need to update this to include the logic -
1. Either it should have types - proposer, canceller, bypasser,
2. Or it should have type- multichain with different roles

MStreet3 replied:
I just removed this, it's not clear why we are asserting which contracts are in the address map.  the docstring says that we should expect nil state values.

MStreet3 committed:
adds error for missing contracts

MStreet3 committed:
refactor(common/changeset): removes bundle check

Removes check that the addresses contains all the
types and versions defined in a bundle.

It is not required that the MCMS state contracts
be non-nil.  There is a Validate() method and the
docstring states that nil values are possible.

MStreet3 committed:
chore: adds unit test on GetContractSet

MStreet3 committed:
fix(solana): removes unnecessary bundle check

MStreet3 committed:
fix(common): removes unnecessary test

cedric-cordenier commented on deployment/common/changeset/state.go:
```diff
@@ -79,78 +79,77 @@ func MaybeLoadMCMSWithTimelockState(env deployment.Environment, chainSelectors [
+	for address, tv := range addresses {
 		switch {
-		case tvStr.Type == timelock.Type && tvStr.Version.String() == timelock.Version.String():
+		case tv.Type == timelock.Type && tv.Version.String() == timelock.Version.String():
```
@MStreet3 Do we not need to check the label here and everywhere below?

MStreet3 committed:
chore: fix tests

MStreet3 committed:
refactor: renames function

MStreet3 committed:
refactor: cleans up implementation

MStreet3 committed:
Merge branch 'develop' into feat/filter-labeled-addresses

MStreet3 committed:
adds changeset

MStreet3 committed:
adds mcms state test

eutopian commented on deployment/address_book.go:
```diff
@@ -315,28 +315,37 @@ func tvKey(tv TypeAndVersion) typeVersionKey {
+// EnsureDeduped ensures that each contract in the bundle only appears once
+// in the address map.  It returns an error if there are more than one instance of a contract.
+// Returns true if every value in the bundle is found once, false otherwise.
+func EnsureDeduped(addrs map[string]TypeAndVersion, bundle []TypeAndVersion) (bool, error) {
```
nice!

MStreet3 committed:
fix: prevents multiple labeled MCMS contracts

MStreet3 committed:
fix flakey test of slice
